### PR TITLE
Validate and protect corrupted block content

### LINKED ISSUE #1735: Handle unrecognized blocks without modifying their markup
While developing a demo plugin for #1619 and renaming blocks as their desired implementation becomes more clear, the user experience of blocks that are no longer recognized is quite bad.  They are turned into `core/freeform` blocks on saving, without any indication of what is going on.

We need to find a way to show in the UI that an "unrecognized" block was detected, and avoid modifying it.

### LINKED ISSUE #1736: Handle invalid blocks without modifying their markup
We need to detect blocks with invalid markup (e.g. added classnames, added or removed tags).

One way to do this (idea from @mtias): Detect "invalid" blocks by re-serializing blocks that we see in the editor.  If it changes beyond a certain threshold (e.g. whitespace doesn't matter), then show the block as a `core/html` block and highlight how the markup would be changed.

Like #1735 it is important that the default "save" operation not alter the markup inside the block.

### PR DESCRIPTION
**Pull request by aduth**
Closes #1735
Closes #1736
Related: #391

This pull request seeks to validate blocks parsed from post content. It does so by regenerating the expected save output upon the initial parse and comparing it against the actual content. If there is a mismatch, the block is flagged as invalid and the user is barred from editing the block (aside from moving or deleting) with a visual indicator of the invalid state:

Invalid block

__Implementation notes:__

- Yes, there's likely a performance impact of creating an initial serialization of the block
- Yes, there's likely false positives that can occur, particular on order of attributes or class names (which in reality should not have an impact on behavior of content). To reduce false positives, both contents are run through the same HTML beautifier we use for serialization. False positives from reordering attributes seems acceptable to me, as we should discourage manual markup modifications, or at least transition the user to a "manual mode". Enhancements could be made here. Editing block markup in Text mode should be unaffected.

__Testing instructions:__

I've left the initial Cover Image block in Demo content as invalid, because I'd like to revisit whether we really want [nested `section` tags](https://github.com/WordPress/gutenberg/blob/6a8e2485675b20b871f9178b51f116176347eab3/blocks/library/cover-image/index.js#L157-L158) in the serialized output. Verify that the first block in post content is flagged as invalid and cannot be edited.

1. Navigate to Gutenberg > Demo
2. Note that the first block has a "Corrupt" warning like the one shown above

### CODE REVIEW

**Comment by afercia** [comment]
Worth noting that, while clicking the block is disabled, using a keyboard it's still possible to focus the cover image title and edit it. Tabbing backwards from the title it's also possible to focus the formatting toolbar buttons and use them:

screen shot 2017-07-17 at 23 31 11

**Comment by aduth** [comment]
>using a keyboard it's still possible to focus the cover image title and edit it

Great point, and a challenging problem to solve if we want to keep the edit render visible in the background (a nice effect I think). A few possibilities that come to mind:

- Intercept the focus event within the block and "skip" to the next valid focus-able element
- Query for all editable inputs within the block after render and disable them
- Wait for something to come of an [`intertness`](https://html.spec.whatwg.org/multipage/interaction.html#inert-subtrees) property, or simulate effect with a component wrapper
- Show only the warning notice, no edit preview

Other ideas welcome.

File: editor/modes/visual-editor/invalid-block-warning.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+ * WordPress dependencies
+ */
+import { Dashicon } from 'components';
+import { __ } from 'i18n';
+
+const warning = (
+	<div className="editor-visual-editor__invalid-block-warning">
+		<Dashicon icon="warning" size={ 40 } />
+		<p>{ __(
+			'Uh oh! This block is corrupt and has been locked to protect ' +
```
Comment (mtias) [inline editor/modes/visual-editor/invalid-block-warning.js]: "This block has been modified externally and has been locked to protect against content loss." cc @jasmussen

**Comment by mtias** [comment]
We should consider adding these buttons:

`[ Convert to Freeform block ] [ Overwrite changes ] [ Do nothing ]`

File: editor/modes/visual-editor/invalid-block-warning.js (hunk 1 above)
Comment (jasmussen) [inline editor/modes/visual-editor/invalid-block-warning.js]: Yep, that sounds good!

**Comment by jasmussen** [comment]
> We should consider adding these buttons:
> [ Convert to Freeform block ] [ Overwrite changes ] [ Do nothing ]

I dig that. Though it beckons _yet another name_ than "Classic Text" for the fallback block. I mean, I do like the "Classic Text" name, I think it's good for when you mean to insert it via the Inserter, and it's more descriptive than "Freeform", but in this context it's obvious that it can contain more than just "text". Should we simply call it "Classic Editor"?

**Comment by mtias** [comment]
Yes, any name works for me. Maybe "convert to generic block"? Also we could offer the option to handle as HTML block.

**Comment by mtias** [comment]
This looks like a great start to me. @nylen what do you think of measuring performance impact with and without validation on a long-ish post? We may consider making this something you can disable via config or filter.

**Comment by mtias** [comment]
@aduth I'm thinking it'd be cool to be able to switch between "old" and "new" renders to visually diff. It could be two tabs next to the warning message.

**Comment by aduth** [comment]
>I'm thinking it'd be cool to be able to switch between "old" and "new" renders to visually diff. It could be two tabs next to the warning message.

Should we do this in a follow-up pull request?

**Comment by mtias** [comment]
> Should we do this in a follow-up pull request?

Yes, just wondered if it could help with the keyboard issues, because we could try not rendering the default view.

**Comment by aduth** [comment]
The issue is that if preview uses `edit` rendering, it's cumbersome to find and disable all inputs within that render.

Alternatively, we could render the `save` representation for corrupt blocks, which unlike the Edit form should include no editable fields, and is arguably more appropriate to show in case "corrupt" means "user made their own edits that may appear in preview".

**Comment by mtias** [comment]
Yes, I was thinking of rendering `save` and `raw` as comparison, not edit.

**Comment by aduth** [comment]
5328e23 shows the `save` representation as the preview, which shouldn't have any editable fields, and avoids the need for any special tab behavior (well... now that I think of it, if the block is a contact form, maybe not so much 😬 ).

**Comment by mtias** [comment]
If it's a contact form, it is likely the save method is `null`.

**Comment by aduth** [comment]
Sure, it's a rare case that the save representation have focusable fields, but still technically possible and not very well handled here.

**Comment by jasmussen** [comment]
I took the liberty of pushing a little visual polish:

screen shot 2017-07-26 at 09 23 27

Matías points about some action buttons seems really worth addressing, though okay to do it in a separate PR. In the mean time, the message feels like it should be rephrased to be less "this is what's up" and more like "this is what it means". I'm not entirely sure how to phrase this without being overly verbose, as it's a rather complex thing to explain. 

Perhaps this begs general help text in some way, or like linking to external documentation. A simple "What does this mean?" link at the end could address it.

**Comment by aduth** [comment]
Since this touches many different files, I don't want to let this linger pending finer details. I assume we'll need to do some revisions in subsequent pull requests, notably of course fixing the "corrupt" demo Cover Image block.

In the meantime, I've rebased and force-pushed with intent to merge.

One additional change I made is allow the root node block list item to be focused so that it can be "selected", but only for arrangement / deletion (no toolbar shown). See feb105e.

**Comment by aduth** [comment]
Actually, I did find one remaining item that should probably be addressed here: If the user tries to edit an invalid block in Text mode, those user edits will not be respected on save.

**Comment by aduth** [comment]
I believe the behavior described in https://github.com/WordPress/gutenberg/pull/1929#issuecomment-318109060 is actually working correctly, and the issue I'm seeing is caused by the wrong URL being assigned when saving a post from the Demo.

Should be: `?page=gutenberg&post_id=568`
Actually: `?page=gutenberg-demo&post_id=568`

Likely a regression introduced in #1967, where `$is_demo` check no longer accounts for whether a post ID is present in query parameters. Will open a separate pull request shortly.

**Comment by aduth** [comment]
I've verified that editing an invalid block in Text mode works as expected: Changes will be respected, but switching back to Visual will still show the block as invalid so long as its reserialization doesn't match.

See also #2032 for demo content save fix.

File: blocks/test/fixtures/core__gallery.json [hunk 2]
```diff
@@ -37,6 +37,8 @@
                     "alt": "title"
                 }
             ]
-        }
+        },
+        "isValid": false,
+        "originalContent": "<div class=\"wp-block-gallery\">\n\t<figure class=\"blocks-gallery-image\">\n\t\t<img src=\"https://cldup.com/uuUqE_dXzy.jpg\" alt=\"title\" />\n\t</figure>\n\t<figure class=\"blocks-gallery-image\">\n\t\t<img src=\"http://google.com/hi.png\" alt=\"title\" />\n\t</figure>\n</div>"
```
Comment (nylen) [inline blocks/test/fixtures/core__gallery.json]: This needs to be fixed.

File: blocks/test/fixtures/core__image__center-caption.json [hunk 3]
```diff
@@ -8,6 +8,8 @@
             "caption": [
                 "Give it a try. Press the \"really wide\" button on the image toolbar."
             ]
-        }
+        },
+        "isValid": false,
+        "originalContent": "<figure class=\"wp-block-image\"><img src=\"https://cldup.com/YLYhpou2oq.jpg\" class=\"aligncenter\"/><figcaption>Give it a try. Press the &quot;really wide&quot; button on the image toolbar.</figcaption></figure>"
```
Comment (nylen) [inline blocks/test/fixtures/core__image__center-caption.json]: This needs to be fixed.

File: blocks/test/fixtures/core__list__ul.json [hunk 4]
```diff
@@ -37,6 +37,8 @@
                     ]
                 }
             ]
-        }
+        },
+        "isValid": false,
+        "originalContent": "<ul><li>Text & Headings</li><li>Images & Videos</li><li>Galleries</li><li>Embeds, like YouTube, Tweets, or other WordPress posts.</li><li>Layout blocks, like Buttons, Hero Images, Separators, etc.</li><li>And <em>Lists</em> like this one of course :)</li></ul>"
```
Comment (nylen) [inline blocks/test/fixtures/core__list__ul.json]: This needs to be fixed.

File: blocks/test/fixtures/core__preformatted.json [hunk 5]
```diff
@@ -15,6 +15,8 @@
                 },
                 "And more!"
             ]
-        }
+        },
+        "isValid": false,
+        "originalContent": "<pre class=\"wp-block-preformatted\">Some <em>preformatted</em> text...<br>And more!</pre>"
```
Comment (nylen) [inline blocks/test/fixtures/core__preformatted.json]: This needs to be fixed.

File: blocks/test/fixtures/core__pullquote.json [hunk 6]
```diff
@@ -11,6 +11,8 @@
             "citation": [
                 "...with a caption"
             ]
-        }
+        },
+        "isValid": false,
+        "originalContent": "<blockquote class=\"wp-block-pullquote\">\n<p>Testing pullquote block...</p><footer>...with a caption</footer>\n</blockquote>"
```
Comment (nylen) [inline blocks/test/fixtures/core__pullquote.json]: This needs to be fixed.

File: blocks/test/fixtures/core__table.json [hunk 7]
```diff
@@ -194,6 +194,8 @@
                     ]
                 }
             ]
-        }
+        },
+        "isValid": false,
+        "originalContent": "<table><thead><tr><th>Version</th><th>Musician</th><th>Date</th></tr></thead><tbody><tr><td><a href=\"https://wordpress.org/news/2003/05/wordpress-now-available/\">.70</a></td><td>No musician chosen.</td><td>May 27, 2003</td></tr><tr><td><a href=\"https://wordpress.org/news/2004/01/wordpress-10/\">1.0</a></td><td>Miles Davis</td><td>January 3, 2004</td></tr><tr><td>Lots of versions skipped, see <a href=\"https://codex.wordpress.org/WordPress_Versions\">the full list</a></td><td>&hellip;</td><td>&hellip;</td></tr><tr><td><a href=\"https://wordpress.org/news/2015/12/clifford/\">4.4</a></td><td>Clifford Brown</td><td>December 8, 2015</td></tr><tr><td><a href=\"https://wordpress.org/news/2016/04/coleman/\">4.5</a></td><td>Coleman Hawkins</td><td>April 12, 2016</td></tr><tr><td><a href=\"https://wordpress.org/news/2016/08/pepper/\">4.6</a></td><td>Pepper Adams</td><td>August 16, 2016</td></tr><tr><td><a href=\"https://wordpress.org/news/2016/12/vaughan/\">4.7</a></td><td>Sarah Vaughan</td><td>December 6, 2016</td></tr></tbody></table>"
```
Comment (nylen) [inline blocks/test/fixtures/core__table.json]: This needs to be fixed.

File: blocks/test/fixtures/core__verse.json [hunk 8]
```diff
@@ -15,6 +15,8 @@
                 },
                 "And more!"
             ]
-        }
+        },
+        "isValid": false,
+        "originalContent": "<pre class=\"wp-block-verse\">A <em>verse</em>…<br>And more!</pre>"
```
Comment (nylen) [inline blocks/test/fixtures/core__verse.json]: This needs to be fixed.

**Comment by nylen** [comment]
There are now several examples of block fixtures with `"isValid": false` in the test suite.  All of these should probably be valid, and we need to add test cases for the ways we expect this feature to detect invalid blocks.

**Comment by aduth** [comment]
See #2038 for updated Cover Image implementation to resolve lingering warning on Demo content screen.

**Pull request merged**